p4-16/psa/examples/psa-example-incremental-checksum.p4

```diff
@@ -0,0 +1,202 @@
+                         inout metadata user_meta,
+                         in psa_ingress_parser_input_metadata_t istd,
+                         out psa_parser_output_metadata_t ostd)
+{
+    state start {
+        buffer.extract(parsed_hdr.ethernet);
+        transition select(parsed_hdr.ethernet.etherType) {
+            0x0800: parse_ipv4;
+            default: accept;
+        }
+    }
+    state parse_ipv4 {
+        buffer.extract(parsed_hdr.ipv4);
+        verify(parsed_hdr.ipv4.ihl == 5, error.UnhandledIPv4Options);
+        transition select(parsed_hdr.ipv4.protocol) {
+            6: parse_tcp;
+            default: accept;
+        }
+    }
+    state parse_tcp {
+        buffer.extract(parsed_hdr.tcp);
+        transition accept;
+    }
+}
+
+// BEGIN:Incremental_Checksum_Table    
+control ingress(inout headers hdr,
+                inout metadata user_meta,
+                PacketReplicationEngine pre,
+                in  psa_ingress_input_metadata_t  istd,
+                out psa_ingress_output_metadata_t ostd) {
+    action drop() {
+      ingress_drop(ostd);
+    }
+    action forward(PortId_t port, bit<32> srcAddr) {
+      user_meta.fwd_metadata.old_srcAddr = hdr.ipv4.srcAddr;
+      hdr.ipv4.srcAddr = srcAddr;
+      send_to_port(ostd, port);      
+    }
+    table route {
+        key = { hdr.ipv4.dstAddr : lpm; }
+        actions = {
+          forward;
+          drop;
+        }
+    }
+    apply {
+        if(hdr.ipv4.isValid()) {
+          route.apply();
+        }
+    }
+}
+// END:Incremental_Checksum_Table        
+
+parser EgressParserImpl(packet_in buffer,
+                        out headers parsed_hdr,
+                        inout metadata user_meta,
+                        in psa_egress_parser_input_metadata_t istd,
+                        out psa_parser_output_metadata_t ostd)
+{
+    state start {
+        transition accept;
+    }
+}
+
+control egress(inout headers hdr,
+               inout metadata user_meta,
+               BufferingQueueingEngine bqe,
+               in  psa_egress_input_metadata_t  istd,
+               out psa_egress_output_metadata_t ostd)
+{
+    apply { }
+}
+
+// BEGIN:Incremental_Checksum_Example
+control DeparserImpl(packet_out packet, inout headers hdr, in metadata user_meta) {
+    InternetChecksum() ck;
+    bit<16> hdrChecksum;
+    apply {
+        // Update IPv4 checksum
+        ck.clear();
+        ck.update({ hdr.ipv4.version,
+            hdr.ipv4.ihl,
+            hdr.ipv4.diffserv,
+            hdr.ipv4.totalLen,
+            hdr.ipv4.identification,
+            hdr.ipv4.flags,
+            hdr.ipv4.fragOffset,
+            hdr.ipv4.ttl,
+            hdr.ipv4.protocol,
+            //hdr.ipv4.hdrChecksum, // intentionally leave this out
+            hdr.ipv4.srcAddr,
+            hdr.ipv4.dstAddr });
+        hdrChecksum = ck.get();
+        // Update TCP checksum
+       ck.clear();
+       ck.remove(hdr.tcp.checksum);
```

Comment:
Nate, you asked me in email earlier whether remove() is correct to do here, and at the time I wasn't sure.  I am pretty darned sure now that this is a correct way to do an incremental TCP checksum now, primarily due to RFC 1624 and my own writing and thinking on the properties of the one's complement checksum.

Reply:
@jafingerhut ,

While this example is totally valid, it is probably not the way incremental checksums are computed in practice.

Instead, the following happens:
1. In the parser you compute the value of payload checksum by removing all the fields that TCP checksum is computed over from hdr.tcp.checksum and store it in the metadata, e.g. meta.tcpPayloadChecksum.
2. In the control, some of the fields that participate in tcp checksum are modified (which is the whole reason to recompute the checksum in the first place)
3. In the deparser, the TCP checksum is initialized with meta.tcpPayloadChecksum and then all the fields that TCP checksum is computed over are added to that checksum. Finally you do ck.get() and store it in hdr.tcp.checksum and then hdr.tcp is emit()ed.

As an optimization, one can remove/add only the fields that are actually modified in step 2.

Reply:
@vgurevich So a separate parser state for parsing a TCP header after IPv4 vs. parsing a TCP header after IPv6, if both IPv4 and IPv6 are supported in a P4 program, then?  Either that, or you need a single parser state that has an 'if' condition that distinguishes those two cases, and I'm not sure whether you'd be interested in mandating support for if statements inside of a parser state.

Reply:
@jafingerhut ,  Yes, something along those lines (unless we want to revisit the separate ChecksumVerification control :) ). If() statement is more readable, but it is way more dangerous. 

Another option is to instantiate the extern outside (of any parser state) and then remove() IPv4/IPv6 fields in the corresponding states whether or not there is a TCP/UDP header afterwards. Then in the TCP parsing state, you add() tcp.CheckSum and remove() the rest of tcp fields. Same for UDP.

It's all not very pretty, but can probably be partially blamed on the quirkiness of the original (i.e. TCP/UDP) design.....

We can try to write a couple of examples and see which one is more palatable.